File: sdks/unity/Assets/AfricaPaymentsMcp/Scripts/Models/PaymentRequest.cs
using System;
using System.Collections.Generic;
using UnityEngine;

namespace AfricaPaymentsMcp.Models
{
    /// <summary>
    /// Request to initiate a payment
    /// </summary>
    [Serializable]
    public class PaymentRequest
    {
        [Tooltip("Payment amount")]
        [SerializeField] private double amount;
        
        [Tooltip("Currency code (e.g., KES, NGN)")]
        [SerializeField] private string currency = "KES";
        
        [Tooltip("Customer phone number")]
        [SerializeField] private string phoneNumber;
        
        [Tooltip("Unique order reference")]
        [SerializeField] private string reference;
        
        [Tooltip("Optional payment description")]
        [SerializeField] private string description;
        
        [Tooltip("Optional callback URL for webhooks")]
        [SerializeField] private string callbackUrl;
        
        [Tooltip("Optional metadata")]
        [SerializeField] private List<MetadataItem> metadata;

        public double Amount 
        { 
            get => amount; 
            set => amount = value; 
        }
        
        public string Currency 
        { 
            get => currency; 
            set => currency = value; 
        }
        
        public string PhoneNumber 
        { 
            get => phoneNumber; 
            set => phoneNumber = value; 
        }
        
        public string Reference 
        { 
            get => reference; 
            set => reference = value; 
        }
        
        public string Description 
        { 
            get => description; 
            set => description = value; 
        }
        
        public string CallbackUrl 
        { 
            get => callbackUrl; 
            set => callbackUrl = value; 
        }
        
        public List<MetadataItem> Metadata 
        { 
            get => metadata; 
            set => metadata = value; 
        }

        /// <summary>
        /// Gets a clean phone number (digits only)
        /// </summary>
        public string CleanPhoneNumber => System.Text.RegularExpressions.Regex.Replace(phoneNumber ?? "", "\\D", "");

        /// <summary>
        /// Validates the request
        /// </summary>
        public bool IsValid => amount > 0 && 
                              !string.IsNullOrEmpty(currency) && 
                              !string.IsNullOrEmpty(phoneNumber) && 
                              !string.IsNullOrEmpty(reference);

        /// <summary>
        /// Converts to JSON for API
        /// </summary>
        public string ToJson()
        {
            var dict = new System.Collections.Generic.Dictionary<string, object>
            {
                ["amount"] = amount,
                ["currency"] = currency,
                ["phoneNumber"] = CleanPhoneNumber,
                ["reference"] = reference
            };

            if (!string.IsNullOrEmpty(description))
                dict["description"] = description;
            
            if (!string.IsNullOrEmpty(callbackUrl))
                dict["callbackUrl"] = callbackUrl;

            return MiniJSON.Json.Serialize(dict);
        }
    }

    /// <summary>
    /// Metadata item for payment requests
    /// </summary>
    [Serializable]
    public class MetadataItem
    {
        public string key;
        public string value;
    }
}

File: sdks/flutter/lib/src/models/payment_request.dart
import 'package:equatable/equatable.dart';
import 'package:json_annotation/json_annotation.dart';

part 'payment_request.g.dart';

/// Request to initiate a payment
@JsonSerializable()
class PaymentRequest extends Equatable {
  /// Payment amount
  final double amount;

  /// Currency code (e.g., KES, NGN, GHS)
  final String currency;

  /// Customer phone number
  final String phoneNumber;

  /// Unique reference for this payment
  final String reference;

  /// Optional description
  final String? description;

  /// Optional callback URL for webhooks
  final String? callbackUrl;

  /// Optional metadata
  final Map<String, dynamic>? metadata;

  const PaymentRequest({
    required this.amount,
    required this.currency,
    required this.phoneNumber,
    required this.reference,
    this.description,
    this.callbackUrl,
    this.metadata,
  });

  factory PaymentRequest.fromJson(Map<String, dynamic> json) =>
      _$PaymentRequestFromJson(json);

  Map<String, dynamic> toJson() => _$PaymentRequestToJson(this);

  PaymentRequest copyWith({
    double? amount,
    String? currency,
    String? phoneNumber,
    String? reference,
    String? description,
    String? callbackUrl,
    Map<String, dynamic>? metadata,
  }) {
    return PaymentRequest(
      amount: amount ?? this.amount,
      currency: currency ?? this.currency,
      phoneNumber: phoneNumber ?? this.phoneNumber,
      reference: reference ?? this.reference,
      description: description ?? this.description,
      callbackUrl: callbackUrl ?? this.callbackUrl,
      metadata: metadata ?? this.metadata,
    );
  }

  @override
  List<Object?> get props => [
        amount,
        currency,
        phoneNumber,
        reference,
        description,
        callbackUrl,
        metadata,
      ];

  /// Validate the request
  bool get isValid {
    return amount > 0 &&
        currency.isNotEmpty &&
        phoneNumber.isNotEmpty &&
        reference.isNotEmpty;
  }

  /// Get cleaned phone number (digits only)
  String get cleanPhoneNumber {
    return phoneNumber.replaceAll(RegExp(r'\D'), '');
  }
}
